#561: Time Zone Foundations

hawkfish opened the pull request:
I think it is the to start committing pieces of this project. This PR includes:
* DuckDB support for the new types
* ICU Time zone settings
* ICU Time zone table function
* ICU `TIMESTAMPTZ`truncation
* ICU `TIMESTAMPTZ` part extraction

There are more functions, data types and even calendar names to support, but I'd prefer to treat them as new features supported by the infrastructure in this PR.

Commit:
Issue #561: Time Zone Support

Add the TimeZone setting for ICU.
Validate new settings.

Commit:
Issue #561: Time Zone Support

Add pg_timezone_names table function.

Commit:
Issue #561: Time Zone Support

Include all time zones, and abbreviate if possible.

Commit:
Issue #561: Time Zone Support

Add temporary icu_part functions.

Commit:
Issue #561: Time Zone Support

Fix spelling of "millennium"

Commit:
Issue #561: Time Zone Support

Implement remaining icu_partxxx functions
and correct the previous ones with adapters.

Commit:
Merge branch 'hawkfish-millennium' into hawkfish-time-zone

Commit:
Merge branch 'hawkfish-millennium' into hawkfish-time-zone

Commit:
Issue #561: Time Zone Support

Read TimeZone setting.
Fix microsecond resolution.

Commit:
Issue #561: Time Zone Support

Implement icu_date_part.

Commit:
Issue #561: Time Zone Support

Move ICU date part functions to a separate file.

Commit:
Issue #561: Time Zone Support

Use separate header for ICU DatePart.

Commit:
Issue #561: Time Zone Support

Update Python build rules.

Commit:
Issue #561: Time Zone Support

Implement ICU DATETRUNC and test it.
Tidy up some clangd stuff.

Commit:
Issue #561: Time Zones

Add the WITH TIME ZONE types.

Commit:
Merge branch 'hawkfish-time-zone' of github.com:hawkfish/duckdb into hawkfish-time-zone

Commit:
Issue #561: Time Zone Support

Merge DuckDB and ICU changes.
Fix clangd whinging.

Commit:
Merge branch 'master' of github.com:duckdb/duckdb into hawkfish-time-zone

Commit:
Issue #561: Time Zone Support

Fix Windows build by fixing some of the inline/DUCKDB_API mess

Commit:
Issue #561: DATETRUNC Time Zone

Extend DATETRUNC to use TIMESTAMP WITH TIME ZONE.

Commit:
Issue #561: DataChunk inlines

Fix the inlining/DUCDB_API errors for DataChunk::size.

Commit:
Issue #561: Catalog AddFunction

Create an API for adding functions with new bindings
to existing catalog entries.

Commit:
Issue #561: DATE_PART Binding

Bind DATE_PART and friends to TIMEZONETZ using the new Catalog::AddFunction API.
Update the tests to reflect the new data types.

Commit:
Issue #561: ICU LAST_DAY

Implement LAST_DATE for TIMEZONETZ.
Change DATE_PART result type to BIGINT to match
non-TZ version.
Fix Epoch offset direction.

Commit:
Merge branch 'hawkfish-time-zone' of github.com:hawkfish/duckdb into hawkfish-time-zone

Commit:
Merge branch 'master' of github.com:duckdb/duckdb into hawkfish-time-zone

Commit:
Issue #561: Binary DATE_PART result

Fix the result type of the binary DATE_PART implementation
to be BIGINT.
Fix the test results for epoch extraction.

hawkfish commented:
Not sure what the TPC-H regression failure is all about - it looks like a script error?

Mytherin commented:
False positive I think:

> Q4 slow (0.17 vs 0.14).

Will re-run, but probably harmless.

Commit:
Issue #561: Time Zone Coverage

Add coverage for TIMESTAMPTZ aggregates that already tested TIMESTAMP.

Mytherin commented on extension/icu/icu-extension.cpp:
```diff
@@ -145,6 +245,20 @@ void ICUExtension::Load(DuckDB &db) {
+	// Time Zones
+	auto &config = DBConfig::GetConfig(*db.instance);
+	config.AddExtensionOption("TimeZone", "The current time zone", LogicalType::VARCHAR, SetICUTimeZone);
+	Value utc("UTC");
```
Does ICU also have the capability of detecting the users' current time zone? Perhaps that would be a more sensible default than UTC.

hawkfish replied:
It would, butI wasn't sure how reliable it was on all platforms. I can change it though so we can get bug reports ;-)

Mytherin replied:
I would hope ICU is at least semi-well tested, considering how widely used it is :)

hawkfish replied:
I meant DuckDB bug reports/feature requests(!) But I'll see what I can do. The one problem is you can't really test it.

hawkfish replied:
Seems to work on my Mac.

Mytherin reviewed (commented):
Thanks for the PR! This looks great.

The main remaining blocker for getting the TZ stuff to work as expected seems to be the casts, which can't be easily overriden/extended from an extension, right?

hawkfish commented on test/sql/types/timestamp/test_timestamp_tz.test:
```diff
@@ -0,0 +1,12 @@
+
+# Cast from string
+query I
+select timestamptz '2021-11-15 02:30:00';
```
Since the casts are just the same as for `TIMESTAMP` but just using a different result type, I don't think there is anything to do here - at least for ISO-8601. But yes, the ability to parse/cast dates in a local format would need hooks. I'm also not sure that we have the tables built for things like `MONTHNAME` (which is not implemented).

Mytherin replied:
I was under the impression timestamptz casts needed access to the list of timezones as well, no? e.g. in Postgres:

```sql
SELECT timestamptz '2021-01-01 19:00:00 US/Pacific';
--  2021-01-02 04:00:00+01
```

hawkfish replied:
Yeah, but as you said, that would require cast extensibility. So the current implementation just uses the existing casting logic.

Mytherin replied:
Future work then :)

hawkfish replied:
Yes! Although I would say that all we really want is `VARCHAR` => `TIMEZONETZ` For the other direction we might want to leave it as it is, but add a `DATE_FORMAT` function.

Commit:
Merge branch 'master' of github.com:duckdb/duckdb into hawkfish-time-zone

Commit:
Issue #561: Default Time Zone

Use the ICU default time zone as the initial setting.
It is not possible to test this in a portable manner,
but you can use:

```sql
SELECT * FROM duckdb_settings() WHERE name = 'TimeZone';
```

to check it manually for your locale.

Note that we use the ID and not the display name,
wich is not defined in the current data set.

Mytherin reviewed (approved):
Thanks for the changes! Looks good to me.
